### PR TITLE
[Security] Add logout configuration for Clear-Site-Data header

### LINKED ISSUE #49266: Enhance security by issuing a Clear-Site-Data header on logout
[`Clear-Site-Data`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Clear-Site-Data#sign_out_of_a_web_site) is a header that we can use to tell browsers to remove some client-side state (see link for details.)

What about sending this header on logout, so that the "Back" button cannot be used to see the logged-in state after a logout?

### PR DESCRIPTION
**Pull request by maxbeckers**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #49266
| License       | MIT
| Doc PR        | symfony/symfony-docs#17900

Enhance security by issuing a Clear-Site-Data header on logout.
* [Clear-Site-Data](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Clear-Site-Data#sign_out_of_a_web_site) Documentation
* Example: https://www.w3.org/TR/clear-site-data/#example-signout

Default config is off.

Config example for all:
```yaml
security:
    # ...
    firewalls:
        main:
            # ...
            logout:
                path: app_logout
                clear_site_data:
                    - "*"
```
Instead of all with the ``*`` it's also possible to add a set of  ``cache``, ``cookies``, ``storage``, ``executionContexts``. For example without cookies it will look like this:
```yaml
security:
    # ...
    firewalls:
        main:
            # ...
            logout:
                path: app_logout
                clear_site_data:
                    - cache
                    - storage
                    - executionContexts
```

**TODO**
- [x] Doc PR symfony/symfony-docs#17900

### CODE REVIEW

File: src/Symfony/Component/Security/Http/EventListener/ClearSiteDataLogoutListener.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * @final
+ */
+class ClearSiteDataLogoutListener implements EventSubscriberInterface
+{
+    private const HEADER_NAME = 'Clear-Site-Data';
+
+    private string $cookieValue;
+
+    /**
+     * @param string $cookieValue The value for the Clear-Site-Data header. Can be '*' or a subset of 'cache', 'cookies', 'storage', 'executionContexts'.
```
Comment (OskarStark) [inline src/Symfony/Component/Security/Http/EventListener/ClearSiteDataLogoutListener.php]:
```suggestion
     * @param string $cookieValue The value for the Clear-Site-Data header.
     *                            Can be '*' or a subset of 'cache', 'cookies', 'storage', 'executionContexts'.
```

File: src/Symfony/Component/Security/Http/EventListener/ClearSiteDataLogoutListener.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    public function __construct(string $cookieValue)
```
Comment (OskarStark) [inline src/Symfony/Component/Security/Http/EventListener/ClearSiteDataLogoutListener.php]:
```suggestion
    public function __construct(private readonly string $cookieValue)
```
?

File: src/Symfony/Component/Security/Http/Tests/EventListener/ClearSiteDataLogoutListenerTest.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+        $listener->onLogout($event);
+
+        $this->assertEquals(++$headerCountBefore, $response->headers->count());
+
+        $this->assertNotNull($response->headers->get('Clear-Site-Data'));
+        $this->assertEquals($clearSiteDataConfig, $response->headers->get('Clear-Site-Data'));
+    }
+
+    public function provideClearSiteDataConfig()
```
Comment (OskarStark) [inline src/Symfony/Component/Security/Http/Tests/EventListener/ClearSiteDataLogoutListenerTest.php]:
```suggestion
    public static function provideClearSiteDataConfig(): iterable
```

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php [hunk 4]
```diff
@@ -251,6 +251,7 @@ private function addFirewallsSection(ArrayNodeDefinition $rootNode, array $facto
                     ->scalarNode('path')->defaultValue('/logout')->end()
                     ->scalarNode('target')->defaultValue('/')->end()
                     ->booleanNode('invalidate_session')->defaultTrue()->end()
+                    ->scalarNode('clear_site_data')->defaultValue('off')->end()
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: why don't we use an array node with enumerated possible values?

Comment (maxbeckers) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: Good idea, didn't know that way, fixed!

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php [hunk 5]
```diff
@@ -251,6 +251,18 @@ private function addFirewallsSection(ArrayNodeDefinition $rootNode, array $facto
...
+                    ->arrayNode('clear_site_data')
+                        ->fixXmlConfig('clear_site_data')
+                        ->performNoDeepMerging()
+                        ->defaultValue(['off'])
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: we don't need the "off" value - either the array is defined, or not

**Review by nicolas-grekas — COMMENTED** [review]
doesn't miss some changes to an xsd file?

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 6]
```diff
@@ -474,6 +473,13 @@ private function createFirewall(ContainerBuilder $container, string $id, array $
                     ->addTag('kernel.event_subscriber', ['dispatcher' => $firewallEventDispatcherId]);
             }
 
+            // add clear site data listener
+            if (!empty($firewall['logout']['clear_site_data']) && !\in_array('off', $firewall['logout']['clear_site_data'])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: `if ($firewall['logout']['clear_site_data'] ?? false) {`

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 7]
```diff
@@ -474,6 +473,13 @@ private function createFirewall(ContainerBuilder $container, string $id, array $
...
+                $container->setDefinition('security.logout.listener.clear_site_data.'.$id, new ChildDefinition('security.logout.listener.clear_site_data'))
+                    ->addArgument(implode(', ', array_map(function (string $clearSiteDataValue) { return '"'.$clearSiteDataValue.'"'; }, $firewall['logout']['clear_site_data'])))
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: `fn ($v) => '"'.$v.'"'`

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php [hunk 8]
```diff
@@ -251,6 +251,18 @@ private function addFirewallsSection(ArrayNodeDefinition $rootNode, array $facto
                     ->scalarNode('path')->defaultValue('/logout')->end()
                     ->scalarNode('target')->defaultValue('/')->end()
                     ->booleanNode('invalidate_session')->defaultTrue()->end()
+                    ->scalarNode('clear_site_data')->defaultValue('off')->end()
```
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: We cannot have both a scalar node and an array node with the same name. That does not make sense.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php [hunk 9]
```diff
@@ -251,6 +251,18 @@ private function addFirewallsSection(ArrayNodeDefinition $rootNode, array $facto
...
+                    ->arrayNode('clear_site_data')
+                        ->fixXmlConfig('clear_site_data')
```
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: This call is on the wrong place. It must go on the parent node (the one that can fix the name of its children)

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php [hunk 10]
```diff
@@ -251,6 +251,18 @@ private function addFirewallsSection(ArrayNodeDefinition $rootNode, array $facto
...
+                        ->performNoDeepMerging()
+                        ->defaultValue(['off'])
+                        ->beforeNormalization()->ifString()->then(function ($v) { return $v ? array_map('trim', explode(',', $v)) : []; })->end()
+                        ->enumPrototype()
+                        ->values([
```
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: should be indented as it is about configuring the prototype node here (the rule is to indent for calls moving to a different builder object and to dedent on the `->end()` call moving back to the parent)

**Comment by maxbeckers** [comment]
Thanks for your reviews @stof and @nicolas-grekas,
yes i missed the xsd file. now i've added as well some tests do verify the new settings. I hope i have now considered all the relevant points.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php [hunk 11]
```diff
@@ -251,6 +251,15 @@ private function addFirewallsSection(ArrayNodeDefinition $rootNode, array $facto
                     ->scalarNode('path')->defaultValue('/logout')->end()
                     ->scalarNode('target')->defaultValue('/')->end()
                     ->booleanNode('invalidate_session')->defaultTrue()->end()
+                    ->arrayNode('clear_site_data')
+                        ->performNoDeepMerging()
+                        ->beforeNormalization()->ifString()->then(function ($v) { return $v ? array_map('trim', explode(',', $v)) : []; })->end()
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: let's use a short closure

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 12]
```diff
@@ -451,8 +451,7 @@ private function createFirewall(ContainerBuilder $container, string $id, array $
                 'logout_path' => $firewall['logout']['path'],
             ]);
 
-            $logoutSuccessListenerId = 'security.logout.listener.default.'.$id;
-            $container->setDefinition($logoutSuccessListenerId, new ChildDefinition('security.logout.listener.default'))
+            $container->setDefinition('security.logout.listener.default.'.$id, new ChildDefinition('security.logout.listener.default'))
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: unrelated change, should be reverted

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 13]
```diff
@@ -474,6 +473,13 @@ private function createFirewall(ContainerBuilder $container, string $id, array $
                     ->addTag('kernel.event_subscriber', ['dispatcher' => $firewallEventDispatcherId]);
             }
 
+            // add clear site data listener
+            if ($firewall['logout']['clear_site_data'] ?? false) {
+                $container->setDefinition('security.logout.listener.clear_site_data.'.$id, new ChildDefinition('security.logout.listener.clear_site_data'))
+                    ->addArgument(implode(', ', array_map(fn ($v) => '"'.$v.'"', $firewall['logout']['clear_site_data'])))
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: processing values at compile time is a bad practice; e.g. it breaks on env vars.
I would pass the array as is here

File: src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTestCase.php [hunk 14]
```diff
@@ -708,6 +709,16 @@ public function testFirewallListenerWithProvider()
         $this->addToAssertionCount(1);
     }
 
+    public function testFirewallLogoutClearSiteData()
+    {
+        $container = $this->getContainer('logout_clear_site_data');
+        $ClearSiteDataConfig = $container->getDefinition('security.firewall.map.config.main')->getArgument(12)['clear_site_data'];
+        $this->assertEquals([
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTestCase.php]: assertSame + on one line

File: src/Symfony/Component/Security/Http/EventListener/ClearSiteDataLogoutListener.php [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+ */
+class ClearSiteDataLogoutListener implements EventSubscriberInterface
+{
+    private const HEADER_NAME = 'Clear-Site-Data';
+
+    /**
+     * @param string $cookieValue The value for the Clear-Site-Data header.
+     *                            Can be '*' or a subset of 'cache', 'cookies', 'storage', 'executionContexts'.
+     */
+    public function __construct(private readonly string $cookieValue)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/EventListener/ClearSiteDataLogoutListener.php]: let's accept an array

**Review by nicolas-grekas — COMMENTED** [review]
Almost good to me :)

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php (hunk 12 above)
Comment (maxbeckers) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: Okay, I'll do that ... was the boy scout rule ...

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php (hunk 13 above)
Comment (maxbeckers) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: Ok

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @maxbeckers.

**Comment by teamflp** [comment]
Awesome
